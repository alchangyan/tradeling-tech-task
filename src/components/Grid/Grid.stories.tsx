import React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';

import Grid, { TGridProps } from './Grid';

const divColors = ['red', 'green', 'blue', 'yellow'];

export default {
  title: 'Components/Grid',
  component: Grid,
  parameters: {
    jest: ['Grid'],
  },
} as Meta;

const Template: Story<TGridProps> = args => (
  <Grid {...args}>
    {divColors.map((bgColor, index) => (
      <div key={index} style={{ backgroundColor: bgColor, height: 80 }}></div>
    ))}
  </Grid>
);

export const Default = Template.bind({});
Default.args = {};

export const Custom = Template.bind({});
Custom.args = {
  pattern: '10% 40% 40% 10%',
  colGap: 10,
  rowGap: 10,
};
