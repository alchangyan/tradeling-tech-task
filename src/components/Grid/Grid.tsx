import * as React from 'react';

import logger from '../../helpers/logger';
import './Grid.scss';

interface TGridProps {
  /**
   * amount of columns
   */
  colCount?: number;
  /**
   * gap between columns
   */
  colGap?: number;
  /**
   * gap between rows
   */
  rowGap?: number;
}

const defaultProps = {
  colCount: 1,
  colGap: 0,
  rowGap: 0,
};

const Grid: React.FC<TGridProps> = ({ colCount, colGap, rowGap, children }) => {
  const gridStyles = React.useMemo(() => {
    if ((colCount || defaultProps.colCount) < 0) {
      logger('Amount of column should be above positive number', 'error')
      return defaultProps;
    }

    return {
      gridTemplateColumns: Array(colCount).fill('auto').join(' '),
      gridColumnGap: colGap || defaultProps.colGap,
      gridRowGap: rowGap || defaultProps.rowGap,
    };
  }, [colCount, colGap, rowGap]);

  return (
    <div className="grid" style={gridStyles}>
      {children}
    </div>
  );
};

Grid.defaultProps = defaultProps;

export type { TGridProps };
export default Grid;
