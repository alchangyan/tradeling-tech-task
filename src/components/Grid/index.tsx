import * as React from 'react';
import Grid, { TGridProps } from './Grid';

export type { TGridProps };
export default React.memo(Grid);
