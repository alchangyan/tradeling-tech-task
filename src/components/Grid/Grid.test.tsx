import React from 'react';
import { render, cleanup, RenderResult } from '@testing-library/react';
import Grid, { TGridProps } from './index';

interface TGetComponentResult {
  wrapper: RenderResult;
  container: HTMLElement,
  comp: HTMLElement,
}

let comp: HTMLElement;
let props: TGridProps = {};
let wrapper: RenderResult;
let container: HTMLElement;

const getComponent = (): TGetComponentResult => {
  wrapper = render(<Grid {...props} />);
  ({container} = wrapper);
  comp = container.firstChild as HTMLElement;

  return {
    wrapper,
    container,
    comp,
  }
};

beforeEach(cleanup);
afterEach(cleanup);

describe('Grid', () => {
  it('renders correctly', () => {
    getComponent();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders childs', () => {
    props = {colCount: 3};
    getComponent();
    expect(comp.style.gridTemplateColumns).toEqual('auto auto auto');
  });
});
