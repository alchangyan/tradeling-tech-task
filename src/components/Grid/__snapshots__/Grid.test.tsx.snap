// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid renders correctly 1`] = `
<div
  class="grid"
  style="grid-template-columns: auto;"
/>
`;
